### PR TITLE
Add tests for basic `private protected` functionality.

### PR DESCRIPTION
**Pull request by gafter**
PR for a first pass of tests added for testing `private protected`

### CODE REVIEW

**Comment by gafter** [comment]
retest windows_debug_vs-integration_prtest please

File: src/Compilers/CSharp/Test/Emit/PrivateProtected.cs [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+    [CompilerTrait(CompilerFeature.PrivateProtected)]
+    public class PrivateProtected : CSharpTestBase
+    {
+        private static readonly string s_keyPairFile = SigningTestHelpers.KeyPairFile;
+        private static readonly string s_publicKeyFile = SigningTestHelpers.PublicKeyFile;
+        private static readonly ImmutableArray<byte> s_publicKey = SigningTestHelpers.PublicKey;
+        private static readonly DesktopStrongNameProvider s_defaultProvider = new SigningTestHelpers.VirtualizedStrongNameProvider(ImmutableArray.Create<string>());
+
+        [Fact]
+        public void RejectIncimpatibleModifiers()
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/PrivateProtected.cs]: typo

**Comment by gafter** [comment]
Follow-up tests under development will appear in https://github.com/gafter/roslyn/pull/22 until a PR on this repository is created, after this PR is merged.

**Comment by gafter** [comment]
test windows_release_unit64_prtest please

**Comment by gafter** [comment]
The failure was

## Regression

**Microsoft.CodeAnalysis.Editor.UnitTests.CodeGeneration.CodeGenerationTests+CSharp.RemoveAttributeWithMultipleAttributes** (from Roslyn.Services.Editor.UnitTests)

Failing for the past 1 build (Since Failed#26 )
Took 0.84 sec.
### Stacktrace

MESSAGE:
System.AggregateException : 3 exception(s) were thrown during test.\r\nNote: exceptions may have been thrown by another test running concurrently with\r\nthis test. This can happen with any tests that share the same ExportProvider.\r\nExamining individual exception stacks may help reveal the original test and source \r\nof the problem.\r\n\r\nException 0:\r\nSystem.NullReferenceException: Object reference not set to an instance of an object.\r\n at Microsoft.VisualStudio.Text.AdornmentLibrary.TextMarkers.Implementation.TextMarkerVisualManager.OnBatchedTagsChanged(Object sender, BatchedTagsChangedEventArgs e)\r\n at Microsoft.VisualStudio.Text.Utilities.GuardedOperations.RaiseEvent[TArgs](Object sender, EventHandler`1 eventHandlers, TArgs args)\r\n\r\nException 1:\r\nSystem.ArgumentException: The specified NormalizedSnapshotSpanCollection extends beyond the end of the TextSnapshot.\r\n at Microsoft.VisualStudio.Text.NormalizedSnapshotSpanCollection..ctor(ITextSnapshot snapshot, IList`1 spans)\r\n at Microsoft.VisualStudio.Text.Projection.Implementation.BufferGraph.MapDownToFirstMatch(SnapshotSpan span, SpanTrackingMode trackingMode, Predicate`1 match)\r\n at Microsoft.VisualStudio.Text.Projection.Implementation.BufferGraph.MapDownToBuffer(SnapshotSpan span, SpanTrackingMode trackingMode, ITextBuffer targetBuffer)\r\n at Microsoft.VisualStudio.Text.Implementation.MappingSpan.GetSpans(ITextBuffer targetBuffer)\r\n at Microsoft.VisualStudio.Text.Implementation.MappingSpan.GetSpans(ITextSnapshot targetSnapshot)\r\n at Microsoft.VisualStudio.Text.InterTextAdornmentSupport.Implementation.InterLineAdornmentManager.OnBatchedTagsChanged(Object sender, BatchedTagsChangedEventArgs e)\r\n at Microsoft.VisualStudio.Text.Utilities.GuardedOperations.RaiseEvent[TArgs](Object sender, EventHandler`1 eventHandlers, TArgs args)\r\n\r\nException 2:\r\nSystem.NullReferenceException: Object reference not set to an instance of an object.\r\n at Microsoft.VisualStudio.Text.InterTextAdornmentSupport.Implementation.InterLineAdornmentManager.PerformLayout(SnapshotPoint trackingPoint)\r\n at Microsoft.VisualStudio.Text.InterTextAdornmentSupport.Implementation.InterLineAdornmentManager.OnBatchedTagsChanged(Object sender, BatchedTagsChangedEventArgs e)\r\n at Microsoft.VisualStudio.Text.Utilities.GuardedOperations.RaiseEvent[TArgs](Object sender, EventHandler`1 eventHandlers, TArgs args)\r\n\r\n\r\n---- System.NullReferenceException : Object reference not set to an instance of an object.\r\n---- System.ArgumentException : The specified NormalizedSnapshotSpanCollection extends beyond the end of the TextSnapshot.\r\n---- System.NullReferenceException : Object reference not set to an instance of an object.
+++++++++++++++++++
STACK TRACE:
at Microsoft.CodeAnalysis.Editor.UnitTests.Workspaces.TestWorkspace.Dispose(Boolean finalize) at Microsoft.CodeAnalysis.Editor.UnitTests.CodeGeneration.CodeGenerationTests.TestContext.Dispose() in q:\roslyn\src\EditorFeatures\Test\CodeGeneration\CodeGenerationTests.cs:line 946 at Microsoft.CodeAnalysis.Editor.UnitTests.CodeGeneration.CodeGenerationTests.<TestRemoveAttributeAsync>d__14`1.MoveNext() in q:\roslyn\src\EditorFeatures\Test\CodeGeneration\CodeGenerationTests.cs:line 519 --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at Microsoft.CodeAnalysis.Editor.UnitTests.CodeGeneration.CodeGenerationTests.CSharp.<RemoveAttributeWithMultipleAttributes>d__83.MoveNext() in q:\roslyn\src\EditorFeatures\Test\CodeGeneration\CodeGenerationTests.CSharp.cs:line 924 --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) ----- Inner Stack Trace #1 (System.NullReferenceException) ----- at Microsoft.VisualStudio.Text.AdornmentLibrary.TextMarkers.Implementation.TextMarkerVisualManager.OnBatchedTagsChanged(Object sender, BatchedTagsChangedEventArgs e) at Microsoft.VisualStudio.Text.Utilities.GuardedOperations.RaiseEvent[TArgs](Object sender, EventHandler`1 eventHandlers, TArgs args) ----- Inner Stack Trace #2 (System.ArgumentException) ----- at Microsoft.VisualStudio.Text.NormalizedSnapshotSpanCollection..ctor(ITextSnapshot snapshot, IList`1 spans) at Microsoft.VisualStudio.Text.Projection.Implementation.BufferGraph.MapDownToFirstMatch(SnapshotSpan span, SpanTrackingMode trackingMode, Predicate`1 match) at Microsoft.VisualStudio.Text.Projection.Implementation.BufferGraph.MapDownToBuffer(SnapshotSpan span, SpanTrackingMode trackingMode, ITextBuffer targetBuffer) at Microsoft.VisualStudio.Text.Implementation.MappingSpan.GetSpans(ITextBuffer targetBuffer) at Microsoft.VisualStudio.Text.Implementation.MappingSpan.GetSpans(ITextSnapshot targetSnapshot) at Microsoft.VisualStudio.Text.InterTextAdornmentSupport.Implementation.InterLineAdornmentManager.OnBatchedTagsChanged(Object sender, BatchedTagsChangedEventArgs e) at Microsoft.VisualStudio.Text.Utilities.GuardedOperations.RaiseEvent[TArgs](Object sender, EventHandler`1 eventHandlers, TArgs args) ----- Inner Stack Trace #3 (System.NullReferenceException) ----- at Microsoft.VisualStudio.Text.InterTextAdornmentSupport.Implementation.InterLineAdornmentManager.PerformLayout(SnapshotPoint trackingPoint) at Microsoft.VisualStudio.Text.InterTextAdornmentSupport.Implementation.InterLineAdornmentManager.OnBatchedTagsChanged(Object sender, BatchedTagsChangedEventArgs e) at Microsoft.VisualStudio.Text.Utilities.GuardedOperations.RaiseEvent[TArgs](Object sender, EventHandler`1 eventHandlers, TArgs args)

File: src/Compilers/CSharp/Test/Emit/PrivateProtected.cs [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+        }
+
+        [Fact]
+        public void PermittedAccessorProtection()
+        {
+            string source =
+@"class Class
+{
+    public int Prop1 { get; private protected set; }
+    protected internal int Prop2 { get; private protected set; }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/PrivateProtected.cs]: `protected internal` is allowed?

File: src/Compilers/CSharp/Test/Emit/PrivateProtected.cs [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+        [Fact]
+        public void RejectIncompatibleModifiers()
+        {
+            string source =
+@"public class Base
+{
+    private internal int Field1;
+    internal private int Field2;
+    private internal protected int Field3;
+    internal protected private int Field4;
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/PrivateProtected.cs]: Include `public`.

File: src/Compilers/CSharp/Test/Emit/PrivateProtected.cs [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+    static private protected int Field1 = 2;
+}
+";
+            CreateStandardCompilation(source, parseOptions: TestOptions.Regular7_2)
+                .VerifyDiagnostics(
+                // (3,34): error CS1057: 'C.Field1': static classes cannot contain protected members
+                //     static private protected int Field1 = 2;
+                Diagnostic(ErrorCode.ERR_ProtectedInStatic, "Field1").WithArguments("C.Field1").WithLocation(3, 34)
+                );
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/PrivateProtected.cs]: Test `private protected` member in `sealed` class. Should generate `CS0628`?

File: src/Compilers/CSharp/Test/Emit/PrivateProtected.cs [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests
+{
+    [CompilerTrait(CompilerFeature.PrivateProtected)]
+    public class PrivateProtected : CSharpTestBase
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/PrivateProtected.cs]: Are there tests for `ISymbol.DeclaredAccessibility`?

File: src/Compilers/CSharp/Test/Emit/PrivateProtected.cs (hunk 3 above)
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/PrivateProtected.cs]: Test modifiers separated by other valid modifiers:
```
private static protected int Field1;
protected partial private int Field2;
```

File: src/Compilers/CSharp/Test/Emit/PrivateProtected.cs (hunk 4 above)
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/PrivateProtected.cs]: Test hiding an `abstract` method.
```
abstract class A
{
    internal abstract void F();
}
abstract class B : A
{
    private protected new void F() { } // CS0533
}
```

Comment (cston) [inline src/Compilers/CSharp/Test/Emit/PrivateProtected.cs]: Test unimplemented inaccessible method.
A.dll:
```
public abstract class A
{
    private protected abstract void F();
}
```
B.dll (no IVT access):
```
class B : A // CS0534
{
}
```

File: src/Compilers/CSharp/Test/Emit/PrivateProtected.cs [hunk 6]
```diff
@@ -0,0 +1,456 @@
...
+        }
+
+        [Fact]
+        public void ForbiddenAccessorProtection_01()
+        {
+            string source =
+@"class Class
+{
+    private protected int Prop1 { get; private protected set; }
+    private int Prop2 { get; private protected set; }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/PrivateProtected.cs]:
```
private protected int Prop3 { get; protected internal set; }
private protected int Prop4 { get; protected set; }
private protected int Prop5 { get; internal set; }
private protected int Prop6 { get; public set; }
```

File: src/Compilers/VisualBasic/Portable/Binding/Binder_Utils.vb [hunk 7]
```diff
@@ -198,6 +199,9 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                     '  Note: Need to be present both MustInherit & NotInheritable for properly reporting #30926
                     foundModifiers = foundModifiers Or currentModifier
                 Else
+                    If currentModifier = SourceMemberFlags.Private Or currentModifier = SourceMemberFlags.Protected Then
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_Utils.vb]: `OrElse`

File: src/Compilers/VisualBasic/Portable/Binding/Binder_Utils.vb [hunk 8]
```diff
@@ -210,6 +214,16 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
+                Dim feature = InternalSyntax.Feature.PrivateProtected
+                Dim requiredVersion = Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.FeatureExtensions.GetLanguageVersion(feature)
+                Dim currentVersion = CType(privateProtectedToken.Parent.SyntaxTree.Options, VisualBasicParseOptions).LanguageVersion
+                If requiredVersion > currentVersion Then
+                    ' The Private Protected feature is not supported in this language version
+                    ReportDiagnostic(diagBag, privateProtectedToken, ERRID.ERR_LanguageVersion,
+                                     currentVersion.GetErrorName(),
+                                     ErrorFactory.ErrorInfo(Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.FeatureExtensions.GetResourceId(feature)),
+                                     New VisualBasicRequiredLanguageVersion(requiredVersion))
+                End If
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_Utils.vb]:
```
InternalSyntax.Parser.CheckFeatureAvailability(
    diagBag,
    privateProtectedToken.GetLocation(),
    DirectCast(privateProtectedToken.SyntaxTree, VisualBasicSyntaxTree).Options.LanguageVersion,
    InternalSyntax.Feature.PrivateProtected)
```

File: src/Compilers/VisualBasic/Test/Emit/PrivateProtected.vb [hunk 9]
```diff
@@ -0,0 +1,567 @@
...
+BC30176: Only one of 'Public', 'Private', 'Protected', 'Friend', or 'Protected Friend' can be specified.
+    Friend Private Field2 As Integer
+           ~~~~~~~
+BC30176: Only one of 'Public', 'Private', 'Protected', 'Friend', or 'Protected Friend' can be specified.
+    Private Friend Protected Field3 As Integer
+            ~~~~~~
+BC30176: Only one of 'Public', 'Private', 'Protected', 'Friend', or 'Protected Friend' can be specified.
+    Friend Protected Private Field4 As Integer
+                     ~~~~~~~
+</errors>)
```
Comment (cston) [inline src/Compilers/VisualBasic/Test/Emit/PrivateProtected.vb]: Test `Private Protected Public` so it's clear we're not just testing existing `Private` + `Protected Friend` behavior.

File: src/Compilers/CSharp/Test/Symbol/Symbols/SymbolErrorTests.cs [hunk 10]
```diff
@@ -3357,10 +3357,22 @@ internal partial struct S { }
...
+                //         internal partial class C { }
+                Diagnostic(ErrorCode.ERR_PartialModifierConflict, "C").WithArguments("NS.A.C").WithLocation(9, 32),
+                // (12,32): error CS0262: Partial declarations of 'A.S' have conflicting accessibility modifiers
+                //         private partial struct S { }
+                Diagnostic(ErrorCode.ERR_PartialModifierConflict, "S").WithArguments("NS.A.S").WithLocation(12, 32)
+                );
+            //var comp = DiagnosticsUtils.VerifyErrorsAndGetCompilationWithMscorlib(text,
+            //    new ErrorDescription { Code = (int)ErrorCode.ERR_PartialModifierConflict, Line = 3, Column = 30 },
+            //    new ErrorDescription { Code = (int)ErrorCode.ERR_PartialModifierConflict, Line = 9, Column = 32 },
+            //    new ErrorDescription { Code = (int)ErrorCode.ERR_PartialModifierConflict, Line = 12, Column = 32 });
```
Comment (cston) [inline src/Compilers/CSharp/Test/Symbol/Symbols/SymbolErrorTests.cs]: Remove commented out code.

**Comment by gafter** [comment]
@cston I've addressed all comments about the existing tests. I've added suggestions for additional tests to https://github.com/dotnet/roslyn/issues/19722 .

Do you have any further comments?

**Pull request merged**